[REVIEW] Clean Up `#include` Dependencies

mdemoret-nv opened the pull request:
Closes #3376

This PR moves a few header files around to fix bad dependencies from `include` -> `src`. Moving forward, the only allowed `#include` direction will be:

- `src` -> `include`
- `src` -> `src_prims`
- `src_prims` -> `include`

To facilitate this, a few changes needed to be made:
1. Functions for the C-API have been separated into their own `*_api.cpp` file (some were combined with C++ files)
2. `host_buffer`, `device_buffer`, `hostAllocator` and `deviceAllocator` were moved from `src_prims` to `include`
3. `#include <common/cumlHandle.hpp>` has been removed from all C++ files.

This PR includes some additional improvements:
- Updated the `include_checker.py` script:
   - Added functionality to check for badly placed `#pragma once`
   - Added functionality to fix some of the existing warnings
   - General refactor to support more checks
- Fixed all incorrect uses of `#pragma once`
- Fixed incorrect uses of `using namespace ...` in header files outside of a namespace
- Removed some unnecessary `#include`

While this touches a lot of files, the actual number of changes is relatively small. Below is a before/after comparison of the include graphs:

**`src/include`:**
Before:
![image](https://user-images.githubusercontent.com/42954918/105561661-ab32fe00-5cd4-11eb-8850-bfeaffaba60f.png)
After:
![image](https://user-images.githubusercontent.com/42954918/105561673-b4bc6600-5cd4-11eb-8bb7-04be91f902b6.png)

**`src/src_prims`:**
Before:
![image](https://user-images.githubusercontent.com/42954918/105561616-79ba3280-5cd4-11eb-8a49-1d0c030df7c1.png)
After:
![image](https://user-images.githubusercontent.com/42954918/105561633-89397b80-5cd4-11eb-9702-27b2a809834b.png)

JohnZed commented on cpp/scripts/include_checker.py:
```diff
@@ -58,62 +107,100 @@ def list_all_source_file(file_regex, srcdirs):
+        warn.append(Issue(False, w, src, line_num))
+
+    if any(
+            reduce(operator.or_,
```
Is the reduce(operator.or_) just equivalent to `any`? Maybe any would be shorter

mdemoret-nv replied:
Fixed

JohnZed commented on cpp/scripts/include_checker.py:
```diff
@@ -58,62 +107,100 @@ def list_all_source_file(file_regex, srcdirs):
         match = IncludeRegex.search(line)
         if match is None:
+            # Check to see if its a pragma once
+            if (not found_pragma_once):
```
using parens around ifs generally not preferred inpython

mdemoret-nv replied:
Fixed.

JohnZed commented on cpp/scripts/include_checker.py:
```diff
@@ -143,43 +231,66 @@ def check_includes_in(src, inplace, top_inc_dirs):
-        errs = []
+    if inplace and len(issues) > 0:
+        had_fixes = False
+        # print("File: {}. Changing lines {}".format(
```
stray comment

mdemoret-nv replied:
Fixed

JohnZed commented on cpp/src/knn/knn_mg.cu:
```diff
@@ -53,36 +52,36 @@ void reduce(Matrix::Data<int64_t> *&out_I, Matrix::floatData_t *&out_D,
-    res_D, k, allocator, stream, internal_streams, n_internal_streams,
-    rowMajorIndex, rowMajorQuery, &start_indices_long);
-}
+// void perform_local_knn(int64_t *res_I, float *res_D,
```
I'm guessing this was for testing?

mdemoret-nv replied:
Actually this symbol was defined multiple times. Should have been removed instead of commented.

JohnZed reviewed (changes requested):
I read the python and cmakelists files in some detail, but went pretty quickly over the headers. Headers look good. Just a couple small changes to check out. Let's discuss timing separately.

teju85 commented on cpp/include/cuml/cluster/dbscan_api.h:
```diff
@@ -17,10 +17,6 @@
 
 #include <cuml/cuml_api.h>
 
-#ifdef __cplusplus
```
Any reason why this ifdef is removed?

mdemoret-nv replied:
This `#ifdef` was removed because the some of the C++ and C-API functions used to be defined in the same file (see `knn.cu`) and would include both the C++ and C-API headers. In this PR I separated all C-API definitions and declarations into `*_api.h/cpp` files which removed the need for `#ifdef __cplusplus`.

I think this better separates the two different compilations and cleans up include tree since C++ files only include C++ headers and vice-versa.

Looking now, `svm_api.h`, `glm_api.h` and `holtwinters_api.h` also have these `#ifdefs` which could be removed as well.

mdemoret-nv replied:
This `#ifdef` was removed because the some of the C++ and C-API functions used to be defined in the same file (see `knn.cu`) and would include both the C++ and C-API headers. In this PR I separated all C-API definitions and declarations into `*_api.h/cpp` files which removed the need for `#ifdef __cplusplus`.

I did accidentally remove the `extern "C"` in other places and fixed that up as a result of this comment. All of the C-API files now follow the same pattern.

teju85 replied:
This header file only contains the 2 dbscan C-API methods. So, shouldn't this header file also be wrapped with the `__cplusplus` macro guard?

mdemoret-nv replied:
I took this into consideration and updated the design to revert these changes. See this comment for more info: https://github.com/rapidsai/cuml/pull/3402#issuecomment-772092988

mdemoret-nv commented on cpp/scripts/include_checker.py:
```diff
@@ -22,32 +22,81 @@
 # file names could (in theory) contain simple white-space
 IncludeRegex = re.compile(r"(\s*#include\s*)([\"<])([\S ]+)([\">])")
+PragmaRegex = re.compile(r"^ *\#pragma\s+once *$")
+UsingNamespaceRegex = re.compile(r"^ *\#pragma\s+once *$")
```
In the PR description I did mention that this PR does add an additional check for bad `#pragma once` usages (not the first line of code).

I was planning on adding a check for poor `using namespace XXX;` but decided against it since there were very little places that benefited and there was already an issue created to address this.

Line 31 should have been removed.

teju85 commented on cpp/CMakeLists.txt:
```diff
@@ -489,10 +498,14 @@ if(BUILD_CUML_C_LIBRARY)
     src/glm/glm_api.cpp
     src/holtwinters/holtwinters_api.cpp
-    src/svm/svm_api.cpp)
+    src/knn/knn_api.cpp
```
It is finally great to see our C-API evolving and getting matured over time!

teju85 commented on cpp/src/knn/knn_api.cpp:
```diff
@@ -0,0 +1,83 @@
+// #include <cuml/common/logger.hpp>
+#include <cuml/neighbors/knn.hpp>
+
+// #include <ml_mg_utils.cuh>
+
+// #include <label/classlabels.cuh>
+// #include <selection/knn.cuh>
+
+// #include <cuda_runtime.h>
+// #include <raft/cuda_utils.cuh>
+
+// #include <sstream>
```
nitpick: Stray comments?

mdemoret-nv replied:
Fixed.

teju85 commented on cpp/src/knn/knn_api.cpp:
```diff
@@ -0,0 +1,83 @@
+#include <cuml/neighbors/knn_api.h>
+#include <common/cumlHandle.hpp>
+
+// #include <cuml/common/logger.hpp>
```
nitpick: stray comments?

mdemoret-nv replied:
Fixed.

teju85 commented on cpp/include/cuml/neighbors/knn_api.h:
```diff
@@ -16,11 +16,8 @@
 
-#include <cuml_api.h>
-
-#ifdef __cplusplus
```
same as dbscan_api.h. Since this is going to be a C-header file, I believe we should keep the `__cplusplus` macro guard to avoid symbol-mangling.

mdemoret-nv replied:
I took this into consideration and updated the design to revert these changes. See this comment for more info: https://github.com/rapidsai/cuml/pull/3402#issuecomment-772092988

teju85 reviewed (changes requested):
It is finally great to see our C-API also evolving!

mdemoret-nv commented:
After reviewing @teju85 comments, I've updated the C-API structure a bit and added back in the `#ifdef __cplusplus` to the C-API headers. I was under the impression we were actually building the C-API with `gcc` which would make these `#ifdef`s moot. However, it looks like we are actually using `g++` and relying on the `extern "C"` to not mangle the names. This was interesting to discover since it was very unlikely any C program could actually use the C-API headers in their previous state.

To ensure the C-API can actually be used, I added a simple test library that is compiled only in C and imports `libcuml.so`. This flushed out a lot of errors that would have prevented anyone from using the C-API.

I would appreciate a re-review on the updated changes if possible. I still believe this PR is low ris because most changes are to `#include` statements which allows us to rely on the compiler to ensure everything is correct.

dantegd commented on cpp/bench/CMakeLists.txt:
```diff
@@ -64,17 +66,18 @@ if(BUILD_CUML_PRIMS_BENCH)
   add_dependencies(prims_benchmark spdlog)
 
-  target_link_libraries(prims_benchmark ${CUDA_cublas_LIBRARY} benchmarklib)
+  target_link_libraries(prims_benchmark ${CUDA_cublas_LIBRARY} benchmarklib ${CUML_CPP_TARGET})
```
perhaps also do one line per entry here as well for consistency

mdemoret-nv replied:
Fixed

dantegd commented on cpp/include/cuml/cuml_api.h:
```diff
@@ -20,13 +20,24 @@
 
 #include <cuda_runtime_api.h>
 
+// Block inclusion of this header when compiling libcuml++.so. If this error is
+// shown during compilation, there is an issue with how the `#include` have
+// been set up. To debug the issue, run `./build.sh cppdocs` and open the page
+// 'cpp/build/html/cuml__api_8h.html' in a browser. This will show which files
+// directly and indirectly include this file. Only files ending in '*_api' or
+// 'cumlHandle' should include this header.
```
This comment might be useful to add to the readme/dev guide (if its already there disrefard comment of course)

mdemoret-nv replied:
Good suggestion. I added this by the error message because I assumed developers would hit the error, go to the source, and then be able to see the message. I'll add it to the dev guide somewhere.

mdemoret-nv replied:
Fixed

dantegd reviewed (approved):
Just left 2 suggestions but changes, particularly the new C api tests look great!

JohnZed commented on cpp/cmake/Dependencies.cmake:
```diff
@@ -39,7 +39,7 @@ else(DEFINED ENV{RAFT_PATH})
   ExternalProject_Add(raft
     GIT_REPOSITORY    https://github.com/rapidsai/raft.git
-    GIT_TAG           9dbf2c8a9134ce8135f7fe947ec523d874fcab6a 
+    GIT_TAG           16a3b8eda8b8c5948d2aad9dd9fcd09d7697e2f3
```
This tag is outdated. Should bring to HEAD: 4a79adcb0c0e87964dcdc9b9122f242b5235b702

mdemoret-nv replied:
Fixed.

JohnZed commented on cpp/include/cuml/cuml_api.h:
```diff
@@ -20,13 +20,24 @@
+// directly and indirectly include this file. Only files ending in '*_api' or
+// 'cumlHandle' should include this header.
+#ifdef CUML_CPP_API
+#error \
```
Nice addition.

JohnZed commented on cpp/src_prims/sparse/distance/l2_distance.cuh:
```diff
@@ -31,11 +31,11 @@
 
-#include <sparse/distance/common.h>
-#include <sparse/distance/ip_distance.cuh>
+#include "common.h"
```
common.h appears to be included twice?

mdemoret-nv replied:
Fixed.

JohnZed reviewed (approved):
Looks great, just one question (and one tiny change). You flipped a couple of includes from absolute to relative and others from relative to absolute. How should we think about the simplest rule for when to prefer absolute or relative #includes?

mdemoret-nv commented:
> Looks great, just one question (and one tiny change). You flipped a couple of includes from absolute to relative and others from relative to absolute. How should we think about the simplest rule for when to prefer absolute or relative #includes?

@JohnZed The changes between relative and absolute were due to the script `include_checker.py` which has an `--inplace` option. Relative includes should only be used if you jump <2 directories in a single direction. Everything else should be absolute. For example:

- `#include ../../my_header.hpp`
   - This is ok because we only jump 2 directories
- `#include ../distance/my_header.hpp`
   - This should be avoided (and will be corrected by the script) because we are going up one and over one. Convert to absolute.

While these rules are the ones the `include_checker.py` script will change, I have been working on a comprehensive set of rules that we should adopt as part of a larger rework of the include structure (that Will and I have been looking at to improve compilation speed). Between the `include_checker.py` script and `clang-format` version 10, almost all of this can be automated and could be run in the style check

mdemoret-nv commented:
rerun tests

teju85 reviewed (approved):
Yes, @mdemoret-nv you are right. For the libraries which expose a C-API, but use `g++` (for whatever reasons), we'll have to put these `__cplusplus` guards. Thank you for fixing it.

Really appreciate you adding a C-API testing in here!

LGTM.

JohnZed commented:
@gpucibot merge
